### PR TITLE
Remove custom red black tree implementation

### PR DESCRIPTION
**Pull request by RonjaPonja**

### CODE REVIEW

File: polysh/display_names.py [hunk 1]
```diff
@@ -16,73 +16,71 @@
...
 max_display_name_length = 0
 
 
-class DisplayNamePrefix(object):
-    def __init__(self):
-        self.next_suffix = 0
-        self.holes = RBTree()
+def acquire_prefix_index(prefix):
+    if prefix not in PREFIXES:
+        PREFIXES[prefix] = list()
```
Comment (hasegeli) [inline polysh/display_names.py]: This smells like you need `defaultdict`.

File: polysh/display_names.py [hunk 2]
```diff
@@ -16,73 +16,71 @@
...
-            prev_suffix_node = self.holes.findNode(prev_suffix)
-            if not prev_suffix_node:
-                return
-            self.holes.deleteNode(prev_suffix_node)
-            self.next_suffix = prev_suffix
 
-    def empty(self):
-        return self.next_suffix == 0
+def release_prefix_index(prev_display_name):
+    split = prev_display_name.split('#')
```
Comment (hasegeli) [inline polysh/display_names.py]: You want to split only once.

File: polysh/display_names.py [hunk 3]
```diff
@@ -16,73 +16,71 @@
...
+    prefix = split[0]
+    if len(split) == 1:
+        suffix = 0
+    else:
+        suffix = int(split[1])
 
+    # We are not deleting the host with the highest suffix. Therefore we need
+    # to mark the current suffix index as unused.
+    if suffix != len(PREFIXES[prefix]) -1:
```
Comment (hasegeli) [inline polysh/display_names.py]: I would write it like `suffix < len()`.

File: polysh/display_names.py [hunk 4]
```diff
@@ -16,73 +16,71 @@
...
+        return
 
-def make_unique_name(prefix):
-    prefix_obj = PREFIXES.get(prefix, None)
-    if prefix_obj is None:
-        prefix_obj = DisplayNamePrefix()
-        PREFIXES[prefix] = prefix_obj
+    # We are deleting the host with thte highest suffix. Therefore we need to
+    # delete it.
+    PREFIXES[prefix].pop(suffix)
```
Comment (hasegeli) [inline polysh/display_names.py]: Maybe `del PREFIXES[prefix][suffix]`

File: polysh/display_names.py [hunk 5]
```diff
@@ -16,73 +16,71 @@
...
+
+    # Remove holes previously left.
+    for idx in reversed(range(len(PREFIXES[prefix]))):
```
Comment (hasegeli) [inline polysh/display_names.py]: I think you need `enumerate` there as well.

File: polysh/display_names.py [hunk 6]
```diff
@@ -16,73 +16,71 @@
...
+        if PREFIXES[prefix][idx]:
+            return
+        PREFIXES[prefix].pop(idx)
```
Comment (hasegeli) [inline polysh/display_names.py]: Again `del` maybe

File: polysh/display_names.py [hunk 7]
```diff
@@ -16,73 +16,71 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from polysh.rb_tree import RBTree
+# The prefix is the key, the value is a list of suffixes in use or None as padding
```
Comment (hasegeli) [inline polysh/display_names.py]: Line too long

File: polysh/display_names.py [hunk 8]
```diff
@@ -16,73 +16,71 @@
...
-PREFIXES = {}
-
-# Red/black tree with key:len(display_name) value:nr of enabled shells with a
+# dict with key:len(display_name) value:nr of enabled shells with a
 # display_name of such a length
-NR_ENABLED_DISPLAY_NAMES_BY_LENGTH = RBTree()
+NR_ENABLED_DISPLAY_NAMES_BY_LENGTH = dict()
 
-# Cache the right most element in the NR_ENABLED_DISPLAY_NAMES_BY_LENGTH tree
+# Used for terminal sizes and layouting
```
Comment (hasegeli) [inline polysh/display_names.py]: "layouting" looks like misspelled.  I would just say "Used for terminal size and layout".

File: polysh/display_names.py [hunk 9]
```diff
@@ -16,73 +16,71 @@
...
-            return
+    # Search and reuse removed host suffix
+    for idx, item in enumerate(PREFIXES[prefix]):
+        if not item:
+            PREFIXES[prefix][idx] = True
+            return idx
+
+    # Add new suffix if no old suffix can be reused
+    PREFIXES[prefix].append(True)
+    return len(PREFIXES[prefix]) -1
```
Comment (hasegeli) [inline polysh/display_names.py]: Missing space after operator

File: polysh/display_names.py [hunk 10]
```diff
@@ -16,73 +16,71 @@
...
+    if suffix != len(PREFIXES[prefix]) -1:
+        PREFIXES[prefix][suffix] = False
+        return
 
-def make_unique_name(prefix):
-    prefix_obj = PREFIXES.get(prefix, None)
-    if prefix_obj is None:
-        prefix_obj = DisplayNamePrefix()
-        PREFIXES[prefix] = prefix_obj
+    # We are deleting the host with thte highest suffix. Therefore we need to
```
Comment (hasegeli) [inline polysh/display_names.py]: `s/thte/the`

File: polysh/display_names.py [hunk 11]
```diff
@@ -16,73 +16,71 @@
...
+def make_unique_name(prefix):
+    suffix = acquire_prefix_index(prefix)
     if suffix:
-        name = '%s#%d' % (prefix, suffix)
+        return '{}#{}'.format(prefix, suffix)
     else:
-        name = prefix
-
-    return name
+        return prefix
```
Comment (hasegeli) [inline polysh/display_names.py]: No need for the `else`

File: polysh/display_names.py [hunk 12]
```diff
@@ -16,73 +16,71 @@
...
+    PREFIXES[prefix].pop(suffix)
+
+    # Remove holes previously left.
+    for idx in reversed(range(len(PREFIXES[prefix]))):
+        if PREFIXES[prefix][idx]:
+            return
+        PREFIXES[prefix].pop(idx)
 
-    suffix = prefix_obj.new_suffix()
+    # If we arrived here, we just deleted the last item with a specific prefix. Therefore we need to delete the whole prefix now.
```
Comment (hasegeli) [inline polysh/display_names.py]: Line too long

File: polysh/display_names.py [hunk 13]
```diff
@@ -16,73 +16,71 @@
...
+        return '{}#{}'.format(prefix, suffix)
     else:
-        name = prefix
-
-    return name
+        return prefix
 
 
 def update_max_display_name_length():
     from polysh import dispatchers
```
Comment (hasegeli) [inline polysh/display_names.py]: Maybe this can be imported properly in the beginning now.

**Review by hasegeli — CHANGES_REQUESTED** [review]

File: polysh/display_names.py (hunk 13 above)
Comment (RonjaPonja) [reply polysh/display_names.py]: Moving it works in python3.5 and python3.6 but not python3.4
https://travis-ci.org/innogames/polysh/jobs/388869052

**Review by hasegeli — APPROVED** [review]

**Pull request merged**